shared/naturalcrit/codeEditor/fold-code.js

```diff
@@ -0,0 +1,338 @@
+			// we can reuse the built-in indicator element if its className matches the new state
+			const clsFolded = typeof opts.indicatorFolded == 'string' && classTest(opts.indicatorFolded);
+			const clsOpen = typeof opts.indicatorOpen == 'string' && classTest(opts.indicatorOpen);
+			cm.eachLine(from, to, function (line) {
+				++cur;
+				let mark = null;
+				let old = line.gutterMarkers;
+				if(old) old = old[opts.gutter];
+				if(isFolded(cm, cur)) {
+					if(clsFolded && old && clsFolded.test(old.className)) return;
+					mark = marker(opts.indicatorFolded);
+				} else {
+					const pos = Pos(cur, 0);
+					const range = func && func(cm, pos);
+					if(range && range.to.line - range.from.line >= minSize) {
+						if(clsOpen && old && clsOpen.test(old.className)) return;
+						mark = marker(opts.indicatorOpen);
+					}
+				}
+				if(!mark && !old) return;
+				cm.setGutterMarker(line, opts.gutter, mark);
+			});
+		};
+
+		// copied from CodeMirror/src/util/dom.js
+		const classTest = function(cls) {
+			return new RegExp(`(^|\\s)${cls}(?:$|\\s)\\s*`);
+		};
+
+		const updateInViewport = function(cm) {
+			const vp = cm.getViewport(), state = cm.state.foldGutter;
+			if(!state) return;
+			cm.operation(function () {
+				updateFoldInfo(cm, vp.from, vp.to);
+			});
+			state.from = vp.from;
+			state.to = vp.to;
+		};
+
+		const onGutterClick = function(cm, line, gutter) {
+			const state = cm.state.foldGutter;
+			if(!state) return;
+			const opts = state.options;
+			if(gutter != opts.gutter) return;
+			const folded = isFolded(cm, line);
+			if(folded) folded.clear();
+			else cm.foldCode(Pos(line, 0), opts);
+		};
+
+		const onChange = function(cm) {
+			const state = cm.state.foldGutter;
+			if(!state) return;
+			const opts = state.options;
+			state.from = state.to = 0;
+			clearTimeout(state.changeUpdate);
+			state.changeUpdate = setTimeout(function () {
+				updateInViewport(cm);
+			}, opts.foldOnChangeTimeSpan || 600);
+		};
+
+		const onViewportChange = function(cm) {
+			const state = cm.state.foldGutter;
+			if(!state) return;
+			const opts = state.options;
+			clearTimeout(state.changeUpdate);
+			state.changeUpdate = setTimeout(function () {
+				const vp = cm.getViewport();
+				if(state.from == state.to || vp.from - state.to > 20 || state.from - vp.to > 20) {
+					updateInViewport(cm);
+				} else {
+					cm.operation(function () {
+						if(vp.from < state.from) {
+							updateFoldInfo(cm, vp.from, state.from);
+							state.from = vp.from;
+						}
+						if(vp.to > state.to) {
+							updateFoldInfo(cm, state.to, vp.to);
+							state.to = vp.to;
+						}
+					});
+				}
+			}, opts.updateViewportTimeSpan || 400);
+		};
+
+		const onFold = function(cm, from) {
+			const state = cm.state.foldGutter;
+			if(!state) return;
+			const line = from.line;
+			if(line >= state.from && line < state.to)
+				updateFoldInfo(cm, line, line + 1);
+		};
+	},
+	registerHomebreweryHelper : function(CodeMirror) {
```

Comment:
Just to add some context to this, I based this function off of the Markdown code folding helper that is packaged with CodeMirror. It will allow for folding the first line in the document to the first `\page`, then the line immediately after each following `\page` to the next `\page`. It will not allow for folding from the last `\page` to the end of the document as it is currently designed.